### PR TITLE
Add datamodule and trainer fixes for validation after each step

### PR DESCRIPTION
**Pull request by SurajPatelM**
Added VideoCaptionDataModule for structured train/val/test loading.

Standardized dataset output format and collate_fn.

Updated trainer to log and validate after each step, saving best model.

Fixed BART decoder to accept BaseModelOutput encoder outputs.

Enables better training monitoring and consistent data handling.

fix: save best model once per epoch

- Move checkpoint write from on_validation_epoch_end to on_train_epoch_end
- Snapshot best weights during validation and save the snapshot at epoch end
- Keeps per-batch validation; prevents N saves per epoch and ensures saved file matches best val loss

### CODE REVIEW

File: data/datamodule.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        super().__init__()
+        self.cfg = cfg
+        self.train_ds = None
+        self.val_ds = None
+        self.test_ds = None
+
+    def _build_dataset(self, split_dir: str, *, use_csv: Optional[bool] = None):
+        use_csv = getattr(self.cfg.data, "use_csv", True) if use_csv is None else use_csv
+        if use_csv:
+            return VideoCaptionDatasetCSV(
```
Comment (vishwajeet-hogale) [inline data/datamodule.py:29]: Let's use the CSVs only for now. But this is good. We can have jsons too.

File: data/datamodule.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            )
+        return VideoCaptionDataset(
+            captions_dir=split_dir,
+            frames_dir=self.cfg.data.frames_dir
+        )
+
+    def setup(self, stage: Optional[str] = None):
+        self.train_ds = self._build_dataset(self.cfg.data.captions_dir)
+        val_dir = getattr(self.cfg.data, "val_captions_dir", None)
+        self.val_ds = self._build_dataset(val_dir or self.cfg.data.captions_dir)
```
Comment (vishwajeet-hogale) [inline data/datamodule.py:41]: So the thing is we don't have a validation dataset yet. I was thinking we'll split NG tube insertion videos into train and test. I think there are too many insertion videos that we can split for validation and test.

File: data/dataset.py [hunk 3]
```diff
@@ -46,7 +46,15 @@ def __getitem__(self, idx):
         image_path, caption = self.data[idx]
         image = Image.open(image_path).convert("RGB")
         image = self.transform(image)
-        return image, caption
+        # Return the same structure as the CSV dataset
+        return {"images": image, "captions": caption}
+
+    @staticmethod
+    def collate_fn(batch):
+        import torch
```
Comment (vishwajeet-hogale) [inline data/dataset.py:54]: Let's move this to the top

File: data/datamodule.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    """
+    def __init__(self, cfg):
+        super().__init__()
+        self.cfg = cfg
+        self.train_ds = None
+        self.val_ds = None
+        self.test_ds = None
+
+    def _build_dataset(self, split_dir: str, *, use_csv: Optional[bool] = None):
+        use_csv = getattr(self.cfg.data, "use_csv", True) if use_csv is None else use_csv
```
Comment (vishwajeet-hogale) [inline data/datamodule.py:27]: Can you add use_csv config param in default.yaml

**Review by vishwajeet-hogale — CHANGES_REQUESTED** [review]
Just reviewed all of this. Very tiny changes. Did you run the code? Does it start training?

File: data/datamodule.py (hunk 2 above)
Comment (SurajPatelM) [reply data/datamodule.py]: Right, I went back to taking it from the original split.

File: data/dataset.py (hunk 3 above)
Comment (SurajPatelM) [reply data/dataset.py]: Removed it and moved everything to the top

File: data/datamodule.py (hunk 4 above)
Comment (SurajPatelM) [reply data/datamodule.py:29]: Included it in the cofig file
